File: algorithms/Python/largestThreeNumbers.py
# O(n) Time | O(1) Space
def findLargestThreeNumbers(array):
    largestThree = [None, None, None]
    for num in array:
        updateLargest(largestThree, num)
    return largestThree

def updateLargest(largestThree, num):
    if largestThree[2] is None or num > largestThree[2]:
        shiftUpdate(largestThree, num, 2)
    elif largestThree[1] is None or num > largestThree[1]:
        shiftUpdate(largestThree, num, 1)
    elif largestThree[0] is None or num > largestThree[0]:
        shiftUpdate(largestThree, num, 0)

def shiftUpdate(array, num, idx):
    # [a, b, c]
    # [0, 1, 2] (indices)
    # for i in range(0, 2 + 1), if i == 2: => [a, b, num]
    for i in range(idx + 1): # python ranges are exclusive
        if i == idx:
            array[i] = num
        else:
            array[i] = array[i + 1]

File: algorithms/Javascript/largestThreeNumbers.js
// O(n) Time | O(1) Space
function findLargestThreeNumbers(array) {
	let largest3Nums = [null, null, null];
	for (const num of array) update(largest3Nums, num);
	return largest3Nums;
}

function update(largest3Nums, num) {
	if (largest3Nums[2] === null || num > largest3Nums[2]) {
		shiftUpdate(largest3Nums, num, 2);
	} else if (largest3Nums[1] === null || num > largest3Nums[1]) {
		shiftUpdate(largest3Nums, num, 1);
	} else if (largest3Nums[0] === null || num > largest3Nums[0]) {
		shiftUpdate(largest3Nums, num, 0);
	}
}

function shiftUpdate(largest3Nums, num, idx) {
	for (let i = 0; i <= idx; i++) {
		if (i === idx) {
			largest3Nums[idx] = num;
		} else {
			largest3Nums[i] = largest3Nums[i + 1];
		}
	}
}